test/polygon_fuzz.cpp

```diff
@@ -0,0 +1,99 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <atomic>
+#include <future>
+
+#include "fuzztest/fuzztest.h"
+#include "gtest/gtest.h"
+#include "polygon.h"
+
+using namespace fuzztest;
+
+void TriangulationNoCrash(
+    std::vector<std::vector<std::pair<float, float>>> input, float precision) {
+  if (precision < 0) precision = -1;
+  manifold::PolygonParams().intermediateChecks = true;
+  manifold::Polygons polys;
+  size_t size = 0;
+  for (const auto &simplePoly : input) {
+    polys.emplace_back();
+    for (const auto &p : simplePoly) {
+      polys.back().emplace_back(p.first, p.second);
+      size++;
+    }
+  }
+  std::atomic<pid_t> tid;
+  std::atomic<bool> faulted(true);
+  auto asyncFuture =
+      std::async(std::launch::async, [&polys, &faulted, &tid, precision]() {
+        tid.store(gettid());
+        try {
+          manifold::Triangulate(polys, precision);
+          faulted.store(false);
+        } catch (manifold::geometryErr e) {
+          // geometryErr is fine
+          faulted.store(false);
+        } catch (...) {
+          printf("got unexpected error\n");
+        }
+      });
+  if (asyncFuture.wait_for(std::chrono::milliseconds(
+          std::max<size_t>(size, 10000))) == std::future_status::timeout) {
+    printf("timeout after %ldms...\n", std::max<size_t>(size, 10000));
+    pthread_cancel(tid.load());
+  }
+
+  EXPECT_FALSE(faulted.load());
+}
+
+using Polygons = std::vector<std::vector<std::pair<float, float>>>;
+using TestCase = std::tuple<Polygons, float>;
+
+std::vector<TestCase> SeedProvider();
+
+FUZZ_TEST(PolygonFuzz, TriangulationNoCrash)
+    .WithDomains(
+        ContainerOf<Polygons>(ContainerOf<std::vector<std::pair<float, float>>>(
+                                  PairOf(Finite<float>(), Finite<float>()))
+                                  .WithMinSize(3)
+                                  .WithMaxSize(10000))
+            .WithMinSize(1)
+            .WithMaxSize(10000),
+        InRange<float>(-1.0, 0.1))
+    .WithSeeds(SeedProvider);
+
+static std::vector<TestCase> TestCases;
+void TestPoly(Polygons polys, int _unused, float precision = -1.0f) {
+  TestCases.push_back({polys, precision});
+}
+
+// ----------------------------------------------------------------------------
+//                        UGLY HACK FOR SEED REUSE
+// ----------------------------------------------------------------------------
+
+#undef TEST
+#define TEST(_unused1, _unused2)
+
+std::vector<TestCase> SeedProvider() {
+  std::vector<TestCase> TestCases;
+  auto TestPoly = [&TestCases](Polygons polys, int _unused,
+                               float precision = -1.0f) {
+    TestCases.push_back({polys, precision});
+  };
+
+#include "polygon_corpus.cpp"
```

Comment:
I'm thinking it might be more useful to make a test or two specific to fuzzing, rather than throwing all of our existing test cases at it. 

Two tests: 
1) overlapping input - ensure manifold and no crash
2) valid input - additionally ensure geometric validity (I expect this one to catch more bugs)

I don't think either of these need to depend on our current suite of specific polygons. For 1, I think effectively random input is fine. For 2, we can ensure validity by taking the same random input from 1 and running it through `CrossSection` to get the positive fill-rule version. 

To get a higher likelihood of generating problem polygons, I'd recommend rounding the coordinates severely to create lots of equality. To additionally test epsilon-valid geometry, the output of `CrossSection` could be perturbed randomly by less than precision.

Reply:
Also, holes are the hardest part, so thinking of ways to tweak the generation such that many holes are formed would also be ideal.

Reply:
The way I understand seed is that seed aids the fuzzer in finding tricky inputs and improve test coverage (e.g. branch coverage). If we only provide one or two seeds, due to the large input space, it can be hard for the fuzzer to get large coverage. Typically, adding more seeds will not harm the fuzz result, but will improve the test result. So I think throwing all our existing test cases is quite useful.

For rounding coordinates, we can probably do https://github.com/google/fuzztest/blob/main/doc/domains-reference.md#reversiblemap-reversible-map. For generating valid geometries, we can define a CSG tree with various primitives, recursively generate the tree, and build the `CrossSection`. I think the CSG tree approach is also helpful for fuzzing our 3D operations. This will probably take a bit of time to do though.

Reply:
I would guess our branch coverage doesn't look too bad anyway, even without any fuzzing. To find new error cases, we need to find new problem situations, which are not that likely to hit a new branch anyway. I'm concerned that by starting with our existing test cases it may actually be harder to find novel problem polygons.

Reply:
I think we can think experiment with this when we cannot find error cases with the fuzzer. It seems too early to talk about this, and I think this is more empirical which requires more experiments to see the result. For now I think it works well.